api/maven-api-xml/src/main/java/org/apache/maven/api/xml/XmlNode.java

```diff
@@ -18,121 +18,244 @@
+    }
+
+    public XmlNode(
+            String prefix,
+            String namespaceUri,
+            String name,
+            String value,
+            Map<String, String> attributes,
+            List<XmlNode> children,
+            Object location) {
+        this.prefix = prefix == null ? "" : prefix;
+        this.namespaceUri = namespaceUri == null ? "" : namespaceUri;
+        this.name = Objects.requireNonNull(name);
+        this.value = value;
+        this.attributes = ImmutableCollections.copy(attributes);
+        this.children = ImmutableCollections.copy(children);
+        this.location = location;
+    }
+    // ----------------------------------------------------------------------
+    // Name handling
+    // ----------------------------------------------------------------------
+
+    public String getPrefix() {
+        return prefix;
+    }
 
-    @Nonnull
-    String getName();
+    public String getNamespaceUri() {
+        return namespaceUri;
+    }
 
-    @Nonnull
-    String getNamespaceUri();
+    public String getName() {
+        return name;
+    }
 
-    @Nonnull
-    String getPrefix();
+    // ----------------------------------------------------------------------
+    // Value handling
+    // ----------------------------------------------------------------------
 
-    @Nullable
-    String getValue();
+    public String getValue() {
+        return value;
+    }
 
-    @Nonnull
-    Map<String, String> getAttributes();
+    // ----------------------------------------------------------------------
+    // Attribute handling
+    // ----------------------------------------------------------------------
 
-    @Nullable
-    String getAttribute(@Nonnull String name);
+    public Map<String, String> getAttributes() {
+        return attributes;
+    }
 
-    @Nonnull
-    List<XmlNode> getChildren();
+    public String getAttribute(String name) {
+        return attributes.get(name);
+    }
 
-    @Nullable
-    XmlNode getChild(String name);
+    // ----------------------------------------------------------------------
+    // Child handling
+    // ----------------------------------------------------------------------
 
-    @Nullable
-    Object getInputLocation();
+    public XmlNode getChild(String name) {
+        if (name != null) {
+            ListIterator<XmlNode> it = children.listIterator(children.size());
+            while (it.hasPrevious()) {
+                XmlNode child = it.previous();
+                if (name.equals(child.getName())) {
+                    return child;
+                }
+            }
+        }
+        return null;
+    }
+
+    public List<XmlNode> getChildren() {
+        return children;
+    }
 
-    default XmlNode merge(@Nullable XmlNode source) {
-        return merge(source, (Boolean) null);
+    public int getChildCount() {
+        return children.size();
     }
 
-    XmlNode merge(@Nullable XmlNode source, @Nullable Boolean childMergeOverride);
```

Comment:
This removes one important piece from the whole API.  It suddenly becomes completely unavailable.  A service needs to be defined.  However, services are currently only available from the `Session`, but it cannot be used here.
A service locator could be used, but that's really overkill.  No better idea at this point though (and don't put the `merge` code in the API, this really needs to be decoupled).